File: backend/src/services/conversationService.js
// src/services/conversationService.js
// Chat Comercial — Comprador ↔ Fornecedor/Prestador, isolado por empresa.
//
// A regra que percorre este ficheiro inteiro: o backend NUNCA confia nos ids
// que o pedido manda para decidir quem pode falar com quem. Uma conversa só
// existe entre duas empresas que uma relação comercial real liga (ou, sem
// contexto, que o próprio pedido authenticado inicia), e só quem pertence a
// uma das duas lê ou escreve nela — ver `conversationComAcesso`, chamada no
// TOPO de toda a operação que toca uma conversa já existente.
const prisma = require('../config/database');
const { NotFoundError, ValidationError, ConflictError } = require('../utils/errors');
const auditService = require('./auditService');
const realtimeService = require('./realtimeService');
const notificationService = require('./notificationService');
const riskAnalysisService = require('./riskAnalysisService');

// Personas de empresa (exclui ADMIN_SISTEMA, que não pertence a nenhuma) —
// reutilizado para notificar "todos na empresa contraparte", tal como
// notifyUsersByRole já faz para outros eventos de negócio.
const PERSONAS_DE_EMPRESA = ['COMPRADOR', 'COMPANY_ADMIN', 'FORNECEDOR', 'FINANCEIRO'];

// Cada tipo de contexto sabe resolver as DUAS empresas que uma conversa
// associada a ele tem de ligar — nunca as que o pedido diz, sempre as que o
// registo em si tem gravadas.
const RESOLVEDORES_DE_CONTEXTO = {
  quote: async (id) => {
    const q = await prisma.quoteRequest.findUnique({ where: { id } });
    if (!q) throw new NotFoundError('Cotação');
    return { buyerCompanyId: q.buyerCompanyId, supplierCompanyId: q.supplierCompanyId };
  },
  purchase_order: async (id) => {
    const po = await prisma.purchaseOrder.findUnique({ where: { id } });
    if (!po) throw new NotFoundError('Ordem de compra');
    return { buyerCompanyId: po.buyerCompanyId, supplierCompanyId: po.supplierCompanyId };
  },
  contract: async (id) => {
    const c = await prisma.contract.findUnique({ where: { id } });
    if (!c) throw new NotFoundError('Contrato');
    return { buyerCompanyId: c.clientCompanyId, supplierCompanyId: c.supplierCompanyId };
  },
  // Um produto só tem fornecedor — o lado comprador é sempre quem está a
  // iniciar a conversa (é para isso que "Falar com fornecedor" existe).
  product: async (id, requesterCompanyId) => {
    const p = await prisma.product.findUnique({ where: { id } });
    if (!p) throw new NotFoundError('Produto');
    return { buyerCompanyId: requesterCompanyId, supplierCompanyId: p.supplierId };
  },
};
const CONTEXT_TYPES = Object.keys(RESOLVEDORES_DE_CONTEXTO);

// Inicia (ou reaproveita) a conversa entre a empresa do utilizador e outra.
//
//  - Com contexto (produto/cotação/pedido/contrato): as DUAS empresas vêm do
//    próprio registo — otherCompanyId, se enviado, é ignorado; o que conta é
//    o que a cotação/PO/contrato/produto realmente liga, e a empresa do
//    utilizador TEM de ser uma das duas, senão é 404 (não revela o resto).
//  - Sem contexto: otherCompanyId é obrigatório — inicia uma conversa direta
//    (ex.: a partir do perfil público de um fornecedor).
async function iniciar({ user, otherCompanyId, contextType, contextId }, req) {
  if (!user.companyId) throw new ValidationError('Só utilizadores de uma empresa podem iniciar o Chat Comercial.');

  let buyerCompanyId;
  let supplierCompanyId;

  if (contextType) {
    if (!CONTEXT_TYPES.includes(contextType) || !contextId) {
      throw new ValidationError('Contexto inválido.');
    }
    const par = await RESOLVEDORES_DE_CONTEXTO[contextType](contextId, user.companyId);
    if (par.buyerCompanyId !== user.companyId && par.supplierCompanyId !== user.companyId) {
      throw new NotFoundError('Contexto');
    }
    ({ buyerCompanyId, supplierCompanyId } = par);
  } else {
    if (!otherCompanyId) throw new ValidationError('Indique a empresa com quem quer falar.');
    if (otherCompanyId === user.companyId) throw new ValidationError('Não pode iniciar uma conversa com a sua própria empresa.');
    const outra = await prisma.company.findUnique({ where: { id: otherCompanyId } });
    if (!outra) throw new NotFoundError('Empresa');
    buyerCompanyId = user.companyType === 'FORNECEDOR' ? otherCompanyId : user.companyId;
    supplierCompanyId = user.companyType === 'FORNECEDOR' ? user.companyId : otherCompanyId;
  }

  const existente = await prisma.conversation.findFirst({
    where: { buyerCompanyId, supplierCompanyId, contextType: contextType || null, contextId: contextId || null, status: 'ABERTA' },
  });
  if (existente) return existente;

  const conversa = await prisma.conversation.create({
    data: { buyerCompanyId, supplierCompanyId, contextType: contextType || null, contextId: contextId || null, createdById: user.id },
  });
  await auditService.recordSafe({
    actor: auditService.actorFrom(req), action: 'CHAT_COMERCIAL_INICIADO',
    entityType: 'Conversation', entityId: conversa.id,
    detail: { ...auditService.contextoFrom(req), buyerCompanyId, supplierCompanyId, contextType: contextType || null, contextId: contextId || null },
  });
  return conversa;
}

async function listarConversas(user) {
  if (!user.companyId) return [];
  const conversas = await prisma.conversation.findMany({
    where: { OR: [{ buyerCompanyId: user.companyId }, { supplierCompanyId: user.companyId }] },
    orderBy: { updatedAt: 'desc' },
  });
  if (!conversas.length) return [];
  const outrasIds = [...new Set(conversas.map((c) => (c.buyerCompanyId === user.companyId ? c.supplierCompanyId : c.buyerCompanyId)))];
  const outras = await prisma.company.findMany({ where: { id: { in: outrasIds } }, select: { id: true, name: true, logoUrl: true } });
  const porId = Object.fromEntries(outras.map((c) => [c.id, c]));
  const ultimaPorConversa = await prisma.conversationMessage.findMany({
    where: { conversationId: { in: conversas.map((c) => c.id) } },
    orderBy: { createdAt: 'desc' }, distinct: ['conversationId'],
  });
  const ultimaPorId = Object.fromEntries(ultimaPorConversa.map((m) => [m.conversationId, m]));
  return conversas.map((c) => ({
    ...c,
    counterpart: porId[c.buyerCompanyId === user.companyId ? c.supplierCompanyId : c.buyerCompanyId] || null,
    lastMessage: ultimaPorId[c.id] || null,
  }));
}

// SÓ participantes — nunca ADMIN_SISTEMA por omissão (ver riskAlertService
// para o acesso do Suporte, que é a EXCEÇÃO, sempre por alerta ativo e sempre
// auditada). 404, não 403: uma conversa entre duas outras empresas não deve
// sequer confirmar que existe para uma terceira.
async function conversationComAcesso(conversationId, user) {
  const conversa = await prisma.conversation.findUnique({ where: { id: conversationId } });
  if (!conversa) throw new NotFoundError('Conversa');
  const participante = user.companyId && (conversa.buyerCompanyId === user.companyId || conversa.supplierCompanyId === user.companyId);
  if (!participante) throw new NotFoundError('Conversa');
  return conversa;
}

// Autorizador do Socket.IO ('conversation:join') — mesma regra, sem lançar.
async function podeEntrarNaConversa(socketUser, conversationId) {
  try {
    await conversationComAcesso(conversationId, socketUser);
    return true;
  } catch {
    return false;
  }
}
realtimeService.setAutorizadores({ conversation: podeEntrarNaConversa });

async function listarMensagens(conversationId, user) {
  await conversationComAcesso(conversationId, user);
  return prisma.conversationMessage.findMany({ where: { conversationId }, orderBy: { createdAt: 'asc' } });
}

async function enviarMensagem(conversationId, user, { body, attachmentUrl, attachmentName }, req) {
  const texto = String(body || '').trim().slice(0, 4000);
  if (!texto && !attachmentUrl) throw new ValidationError('Escreva uma mensagem ou anexe um ficheiro.');

  const conversa = await conversationComAcesso(conversationId, user);
  if (conversa.status === 'FECHADA') throw new ConflictError('Esta conversa está fechada.');

  const mensagem = await prisma.conversationMessage.create({
    data: { conversationId, senderId: user.id, senderCompanyId: user.companyId, body: texto, attachmentUrl, attachmentName },
  });
  await prisma.conversation.update({ where: { id: conversationId }, data: { updatedAt: new Date() } });

  realtimeService.emitToConversation(conversationId, 'conversation:message', mensagem);

  const outraCompanyId = conversa.buyerCompanyId === user.companyId ? conversa.supplierCompanyId : conversa.buyerCompanyId;
  await notificationService.notifyUsersByRole({
    companyId: outraCompanyId, roles: PERSONAS_DE_EMPRESA, type: 'CHAT_COMERCIAL_MENSAGEM', channel: 'IN_APP',
    title: 'Nova mensagem no Chat Comercial', message: `Recebeu uma nova mensagem de ${user.name}.`,
    relatedEntityType: 'Conversation', relatedEntityId: conversationId,
  });

  // Trust & Safety: TODA mensagem é classificada; só MEDIUM+ vira alerta —
  // é assim que "o Suporte não recebe todas as conversas comerciais" se
  // cumpre. Nunca bloqueia (versão atual só regista e avisa o Suporte).
  await analisarRisco(conversa, mensagem, req);

  return mensagem;
}

async function analisarRisco(conversa, mensagem, req) {
  const alertaAberto = await prisma.riskAlert.findFirst({
    where: { conversationId: conversa.id, status: { in: ['ABERTO', 'EM_ANALISE'] }, level: { in: ['MEDIUM', 'HIGH', 'CRITICAL'] } },
  });
  const resultado = riskAnalysisService.analisar(mensagem.body, { escalar: Boolean(alertaAberto) });
  if (resultado.level === 'LOW') return null;

  const alerta = await prisma.riskAlert.create({
    data: {
      conversationId: conversa.id, messageId: mensagem.id, level: resultado.level,
      reason: resultado.reason, signals: resultado.signals,
      context: { buyerCompanyId: conversa.buyerCompanyId, supplierCompanyId: conversa.supplierCompanyId, contextType: conversa.contextType, contextId: conversa.contextId },
    },
  });
  await auditService.recordSafe({
    actor: auditService.actorFrom(req), action: 'ALERTA_SEGURANCA_CRIADO',
    entityType: 'RiskAlert', entityId: alerta.id,
    detail: { ...auditService.contextoFrom(req), conversationId: conversa.id, level: resultado.level, reason: resultado.reason },
  });
  realtimeService.emitToConversation(conversa.id, 'conversation:risk-alert', { level: alerta.level });
  return alerta;
}

async function marcarLidas(conversationId, user) {
  await conversationComAcesso(conversationId, user);
  await prisma.conversationMessage.updateMany({
    where: { conversationId, senderCompanyId: { not: user.companyId }, readAt: null },
    data: { readAt: new Date() },
  });
}

async function contarNaoLidas(user) {
  if (!user.companyId) return 0;
  return prisma.conversationMessage.count({
    where: {
      readAt: null,
      senderCompanyId: { not: user.companyId },
      conversation: { OR: [{ buyerCompanyId: user.companyId }, { supplierCompanyId: user.companyId }] },
    },
  });
}

module.exports = {
  CONTEXT_TYPES, iniciar, listarConversas, conversationComAcesso, listarMensagens,
  enviarMensagem, marcarLidas, contarNaoLidas,
};

File: backend-java/src/main/java/ao/kixima/conversation/ConversationService.java
package ao.kixima.conversation;

import ao.kixima.audit.AuditService;
import ao.kixima.catalog.Product;
import ao.kixima.catalog.ProductRepository;
import ao.kixima.common.error.ConflictException;
import ao.kixima.common.error.NotFoundException;
import ao.kixima.common.error.ServiceUnavailableException;
import ao.kixima.common.error.ValidationException;
import ao.kixima.company.Company;
import ao.kixima.company.CompanyRepository;
import ao.kixima.company.CompanyType;
import ao.kixima.notification.NotificationChannel;
import ao.kixima.notification.NotificationService;
import ao.kixima.notification.NotificationType;
import ao.kixima.po.PurchaseOrder;
import ao.kixima.conversation.dto.ConversationMessageDto;
import ao.kixima.po.PurchaseOrderRepository;
import ao.kixima.quote.QuoteRequest;
import ao.kixima.quote.QuoteRequestRepository;
import ao.kixima.realtime.RealtimeService;
import ao.kixima.risk.RiskAlert;
import ao.kixima.risk.RiskAlertRepository;
import ao.kixima.risk.RiskAlertStatus;
import ao.kixima.risk.RiskAnalysisService;
import ao.kixima.risk.RiskLevel;
import ao.kixima.security.CurrentUser;
import ao.kixima.security.PersonaRole;
import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;

/**
 * Espelha backend/src/services/conversationService.js — Chat Comercial,
 * Comprador ↔ Fornecedor/Prestador, isolado por empresa. Regra central: o
 * backend NUNCA confia nos ids que o pedido manda para decidir quem pode
 * falar com quem — {@link #conversationComAcesso} é chamada no TOPO de
 * toda a operação sobre uma conversa já existente.
 *
 * Contextos "product", "purchase_order" e "quote" resolvem-se contra os
 * domínios já portados. "contract" (Contract) recusa-se explicitamente
 * (503) em vez de fingir suporte — o domínio ainda não existe em Java.
 *
 * Tempo real (M6): {@code realtimeService.emitToConversation} é
 * {@link RealtimeService#emitToConversation} (STOMP, depois do commit) e o
 * autorizador de {@code conversation:join} é {@link #conversationComAcesso},
 * chamado por {@link ao.kixima.realtime.RealtimeAuthInterceptor} no SUBSCRIBE.
 */
@Service
public class ConversationService {

    private static final Set<PersonaRole> PERSONAS_DE_EMPRESA =
            Set.of(PersonaRole.COMPRADOR, PersonaRole.COMPANY_ADMIN, PersonaRole.FORNECEDOR, PersonaRole.FINANCEIRO);
    private static final Set<RiskAlertStatus> ALERTA_ABERTO_STATUS = Set.of(RiskAlertStatus.ABERTO, RiskAlertStatus.EM_ANALISE);
    private static final Set<RiskLevel> ALERTA_ABERTO_NIVEIS = Set.of(RiskLevel.MEDIUM, RiskLevel.HIGH, RiskLevel.CRITICAL);

    private final ConversationRepository conversationRepository;
    private final ConversationMessageRepository messageRepository;
    private final CompanyRepository companyRepository;
    private final ProductRepository productRepository;
    private final PurchaseOrderRepository purchaseOrderRepository;
    private final QuoteRequestRepository quoteRequestRepository;
    private final AuditService auditService;
    private final NotificationService notificationService;
    private final RiskAnalysisService riskAnalysisService;
    private final RiskAlertRepository riskAlertRepository;
    private final ObjectMapper objectMapper;
    private final RealtimeService realtimeService;

    public ConversationService(ConversationRepository conversationRepository, ConversationMessageRepository messageRepository,
                                CompanyRepository companyRepository, ProductRepository productRepository,
                                PurchaseOrderRepository purchaseOrderRepository, QuoteRequestRepository quoteRequestRepository,
                                AuditService auditService,
                                NotificationService notificationService, RiskAnalysisService riskAnalysisService,
                                RiskAlertRepository riskAlertRepository, ObjectMapper objectMapper,
                                RealtimeService realtimeService) {
        this.realtimeService = realtimeService;
        this.conversationRepository = conversationRepository;
        this.messageRepository = messageRepository;
        this.companyRepository = companyRepository;
        this.productRepository = productRepository;
        this.purchaseOrderRepository = purchaseOrderRepository;
        this.quoteRequestRepository = quoteRequestRepository;
        this.auditService = auditService;
        this.notificationService = notificationService;
        this.riskAnalysisService = riskAnalysisService;
        this.riskAlertRepository = riskAlertRepository;
        this.objectMapper = objectMapper;
    }

    private record ParDeEmpresas(String buyerCompanyId, String supplierCompanyId) {
    }

    private ParDeEmpresas resolverContexto(String contextType, String contextId, String requesterCompanyId) {
        return switch (contextType) {
            case "product" -> {
                Product p = productRepository.findById(contextId).orElseThrow(() -> new NotFoundException("Produto"));
                yield new ParDeEmpresas(requesterCompanyId, p.getSupplierId());
            }
            case "purchase_order" -> {
                PurchaseOrder po = purchaseOrderRepository.findById(contextId).orElseThrow(() -> new NotFoundException("Ordem de compra"));
                yield new ParDeEmpresas(po.getBuyerCompanyId(), po.getSupplierCompanyId());
            }
            case "quote" -> {
                QuoteRequest q = quoteRequestRepository.findById(contextId).orElseThrow(() -> new NotFoundException("Cotação"));
                yield new ParDeEmpresas(q.getBuyerCompanyId(), q.getSupplierCompanyId());
            }
            case "contract" -> throw new ServiceUnavailableException(
                    "Conversas associadas a \"contract\" ainda não estão disponíveis neste backend (Java) — "
                            + "dependem do domínio Contract, ainda não portado. Use o backend Node por agora.");
            default -> throw new ValidationException("Contexto inválido.");
        };
    }

    private static final Set<String> CONTEXT_TYPES = Set.of("quote", "purchase_order", "contract", "product");

    @Transactional
    public Conversation iniciar(CurrentUser user, String otherCompanyId, String contextType, String contextId, HttpServletRequest req) {
        if (user.companyId() == null) throw new ValidationException("Só utilizadores de uma empresa podem iniciar o Chat Comercial.");

        String buyerCompanyId;
        String supplierCompanyId;

        if (contextType != null && !contextType.isBlank()) {
            if (!CONTEXT_TYPES.contains(contextType) || contextId == null || contextId.isBlank()) {
                throw new ValidationException("Contexto inválido.");
            }
            ParDeEmpresas par = resolverContexto(contextType, contextId, user.companyId());
            if (!par.buyerCompanyId().equals(user.companyId()) && !par.supplierCompanyId().equals(user.companyId())) {
                throw new NotFoundException("Contexto");
            }
            buyerCompanyId = par.buyerCompanyId();
            supplierCompanyId = par.supplierCompanyId();
        } else {
            if (otherCompanyId == null || otherCompanyId.isBlank()) throw new ValidationException("Indique a empresa com quem quer falar.");
            if (otherCompanyId.equals(user.companyId())) throw new ValidationException("Não pode iniciar uma conversa com a sua própria empresa.");
            Company outra = companyRepository.findById(otherCompanyId).orElseThrow(() -> new NotFoundException("Empresa"));
            boolean euSouFornecedor = user.companyType() == CompanyType.FORNECEDOR;
            buyerCompanyId = euSouFornecedor ? otherCompanyId : user.companyId();
            supplierCompanyId = euSouFornecedor ? user.companyId() : otherCompanyId;
        }

        Conversation existente = conversationRepository
                .findFirstByBuyerCompanyIdAndSupplierCompanyIdAndContextTypeAndContextIdAndStatus(
                        buyerCompanyId, supplierCompanyId, contextType, contextId, ConversationStatus.ABERTA)
                .orElse(null);
        if (existente != null) return existente;

        Conversation conversa = new Conversation(UUID.randomUUID().toString(), buyerCompanyId, supplierCompanyId,
                contextType, contextId, user.id(), Instant.now());
        conversationRepository.save(conversa);

        Map<String, Object> detail = new java.util.LinkedHashMap<>(auditService.contextoFrom(req));
        detail.put("buyerCompanyId", buyerCompanyId);
        detail.put("supplierCompanyId", supplierCompanyId);
        detail.put("contextType", contextType);
        detail.put("contextId", contextId);
        auditService.recordSafe(new AuditService.Entry(auditService.actorFrom(user, req), "CHAT_COMERCIAL_INICIADO",
                "Conversation", conversa.getId(), null, detail));

        return conversa;
    }

    public record ConversaResumo(Conversation conversation, Company counterpart, ConversationMessage lastMessage) {
    }

    @Transactional(readOnly = true)
    public List<ConversaResumo> listarConversas(CurrentUser user) {
        if (user.companyId() == null) return List.of();
        List<Conversation> conversas = conversationRepository
                .findByBuyerCompanyIdOrSupplierCompanyIdOrderByUpdatedAtDesc(user.companyId(), user.companyId());
        if (conversas.isEmpty()) return List.of();

        Set<String> outrasIds = conversas.stream()
                .map(c -> c.getBuyerCompanyId().equals(user.companyId()) ? c.getSupplierCompanyId() : c.getBuyerCompanyId())
                .collect(java.util.stream.Collectors.toSet());
        Map<String, Company> porId = companyRepository.findAllById(outrasIds).stream()
                .collect(java.util.stream.Collectors.toMap(Company::getId, c -> c));

        return conversas.stream().map(c -> {
            Company counterpart = porId.get(c.getBuyerCompanyId().equals(user.companyId()) ? c.getSupplierCompanyId() : c.getBuyerCompanyId());
            ConversationMessage last = messageRepository.findFirstByConversationIdOrderByCreatedAtDesc(c.getId()).orElse(null);
            return new ConversaResumo(c, counterpart, last);
        }).toList();
    }

    /** SÓ participantes — nunca ADMIN_SISTEMA por omissão (ver RiskAlertService para a excepção do Suporte). 404, não 403. */
    @Transactional(readOnly = true)
    public Conversation conversationComAcesso(String conversationId, CurrentUser user) {
        Conversation conversa = conversationRepository.findById(conversationId).orElseThrow(() -> new NotFoundException("Conversa"));
        boolean participante = user.companyId() != null
                && (conversa.getBuyerCompanyId().equals(user.companyId()) || conversa.getSupplierCompanyId().equals(user.companyId()));
        if (!participante) throw new NotFoundException("Conversa");
        return conversa;
    }

    @Transactional(readOnly = true)
    public List<ConversationMessage> listarMensagens(String conversationId, CurrentUser user) {
        conversationComAcesso(conversationId, user);
        return messageRepository.findByConversationIdOrderByCreatedAtAsc(conversationId);
    }

    @Transactional
    public ConversationMessage enviarMensagem(String conversationId, CurrentUser user, String body, String attachmentUrl,
                                               String attachmentName, HttpServletRequest req) {
        String texto = body == null ? "" : body.strip();
        if (texto.length() > 4000) texto = texto.substring(0, 4000);
        if (texto.isEmpty() && (attachmentUrl == null || attachmentUrl.isBlank())) {
            throw new ValidationException("Escreva uma mensagem ou anexe um ficheiro.");
        }

        Conversation conversa = conversationComAcesso(conversationId, user);
        if (conversa.getStatus() == ConversationStatus.FECHADA) throw new ConflictException("Esta conversa está fechada.");

        ConversationMessage mensagem = new ConversationMessage(UUID.randomUUID().toString(), conversationId, user.id(),
                user.companyId(), texto, attachmentUrl, attachmentName, Instant.now());
        messageRepository.save(mensagem);
        conversa.tocarAtualizacao();

        String outraCompanyId = conversa.getBuyerCompanyId().equals(user.companyId()) ? conversa.getSupplierCompanyId() : conversa.getBuyerCompanyId();
        notificationService.notifyUsersByRole(outraCompanyId, List.copyOf(PERSONAS_DE_EMPRESA), NotificationType.CHAT_COMERCIAL_MENSAGEM,
                "Nova mensagem no Chat Comercial", "Recebeu uma nova mensagem de " + user.name() + ".",
                NotificationChannel.IN_APP, "Conversation", conversationId);

        // Trust & Safety: TODA mensagem é classificada; só MEDIUM+ vira alerta. Nunca bloqueia.
        analisarRisco(conversa, mensagem, user, req);

        realtimeService.emitToConversation(conversationId, "conversation:message", ConversationMessageDto.de(mensagem));
        return mensagem;
    }

    private void analisarRisco(Conversation conversa, ConversationMessage mensagem, CurrentUser user, HttpServletRequest req) {
        boolean alertaAberto = riskAlertRepository.findFirstByConversationIdAndStatusInAndLevelInOrderByCreatedAtDesc(
                conversa.getId(), List.copyOf(ALERTA_ABERTO_STATUS), List.copyOf(ALERTA_ABERTO_NIVEIS)).isPresent();
        RiskAnalysisService.Resultado resultado = riskAnalysisService.analisar(mensagem.getBody(), alertaAberto);
        if (resultado.level() == RiskLevel.LOW) return;

        String signalsJson = escreverJsonSilencioso(resultado.signals());
        String contextJson = escreverJsonSilencioso(Map.of(
                "buyerCompanyId", conversa.getBuyerCompanyId(), "supplierCompanyId", conversa.getSupplierCompanyId(),
                "contextType", conversa.getContextType() == null ? "" : conversa.getContextType(),
                "contextId", conversa.getContextId() == null ? "" : conversa.getContextId()));

        RiskAlert alerta = new RiskAlert(UUID.randomUUID().toString(), conversa.getId(), mensagem.getId(), resultado.level(),
                resultado.reason(), signalsJson, contextJson, Instant.now());
        riskAlertRepository.save(alerta);

        Map<String, Object> detail = new java.util.LinkedHashMap<>(auditService.contextoFrom(req));
        detail.put("conversationId", conversa.getId());
        detail.put("level", resultado.level().name());
        detail.put("reason", resultado.reason());
        auditService.recordSafe(new AuditService.Entry(auditService.actorFrom(user, req), "ALERTA_SEGURANCA_CRIADO",
                "RiskAlert", alerta.getId(), null, detail));
        realtimeService.emitToConversation(conversa.getId(), "conversation:risk-alert", Map.of("level", alerta.getLevel().name()));
    }

    private String escreverJsonSilencioso(Object valor) {
        try {
            return objectMapper.writeValueAsString(valor);
        } catch (Exception e) {
            return null;
        }
    }

    @Transactional
    public void marcarLidas(String conversationId, CurrentUser user) {
        conversationComAcesso(conversationId, user);
        messageRepository.marcarLidas(conversationId, user.companyId(), Instant.now());
    }

    @Transactional(readOnly = true)
    public long contarNaoLidas(CurrentUser user) {
        if (user.companyId() == null) return 0;
        return messageRepository.contarNaoLidas(user.companyId());
    }
}
